test/copy.jl

```diff
@@ -253,8 +253,19 @@ end
     @test copyto!(s, String[]) == [1, 2] # No error
 end
 
+function make_circular_reference_array_56775()
+    p = Any[nothing]
+    p[1] = p
+    return p
+end
+
 @testset "deepcopy_internal arrays" begin
     @test (@inferred Base.deepcopy_internal(zeros(), IdDict())) == zeros()
+
+    # issue 56775
+    p = make_circular_reference_array_56775()
+    p2 = deepcopy(p)
+    @test p2 === p2[1]
 end
```

Comment:
I'd put this in a separate testset because it is an end to end test, not a deepcopy_internal test:
```suggestion
@testset "circular reference arrays" begin
    # issue 56775
    p = Any[nothing]
    p[1] = p
    p2 = deepcopy(p)
    @test p2 === p2[1]
end

@testset "deepcopy_internal arrays" begin
    @test (@inferred Base.deepcopy_internal(zeros(), IdDict())) == zeros()
end
```

Reply:
Good point!